File: cpp/leetcode/ConcatenatedWords.cpp
//Leetcode Problem No. 472 Concatenated Words
//Solution written by Xuqiang Fang on 24 May, 2018

#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>
#include <algorithm>

using namespace std;

class Solution{
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        sort(words.begin(), words.end(), com);
        /** debug
        for(auto s : words){
            cout << s << endl;
        }
        */
        vector<string> ans;
        unordered_set<string> dict;
        for(int i=0; i<words.size(); ++i){
            if(can(words[i], dict)){
                ans.push_back(words[i]);
            }
            dict.insert(words[i]);
        }
        return ans;
    }
private:
    /*
    bool compare(const string &a, const string &b){
        return (a.size() < b.size());
    }
    */
    struct compare {
        bool operator()(const std::string& first, const std::string& second) {
            return first.size() < second.size();
        }
    } com;
    int can(string s, unordered_set<string> &d){
        if(d.empty()){
            return 0; 
        }
        int l = s.length();
        vector<int> dp(l+1, 0);
        dp[0] = 1;
        s = " " + s;
        for(int i=1; i<=l; ++i){
            for(int j=0; j<i; ++j){
                if(dp[j] == 1 && d.count(s.substr(j+1, i-j))){
                    dp[i] = 1;
                    break;
                }
            }
        }
        return dp[l];
    }
};

int main(){
    Solution s;
    vector<string> words{"cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"};
    vector<string> ans = s.findAllConcatenatedWordsInADict(words);
    for(auto s : ans){
        cout << s << endl;
    }
    return 0;
}

File: java/leetcode/ConcatenatedWords.java
//Leetcode problem 472 Concatenated Words
//Solution written by Xuqiang Fang on 23 May, 2018

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

class Solution{
    public List<String> findAllConcatenatedWordsInADict(String[] words){
        List<String> ans = new ArrayList<>();
        Arrays.sort(words, new Comparator<String>(){
            public int compare(String a, String b){
                return a.length() - b.length();
            }
        });
        for(String w : words){
            System.out.println(w);
        }
        Set<String> set = new HashSet<>();
        for(int i=0; i<words.length; ++i){
            if(canCon(words[i], set)){
                ans.add(words[i]);
            }
            set.add(words[i]);
        }
        return ans;
    }

    private boolean canCon(String w, Set<String> d){
        if(d.isEmpty()){
            return false;
        }
        boolean[] dp = new boolean[w.length()+1];
        dp[0] = true;
        for(int i=1; i<=w.length(); ++i){
            for(int j=0; j<i; ++j){
                if(!dp[j]){
                    continue;
                }
                if(d.contains(w.substring(j,i))){
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[w.length()];
    }
}

public class ConcatenatedWords{
	public static void main(String[] args){
		Solution s = new Solution();
        String[] words = {"cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"};
        System.out.println(s.findAllConcatenatedWordsInADict(words));
	}
}
